### ISSUE TITLE
Question: Is pydna.assembly deterministic?

### ISSUE THREAD

**Issue by jlerman44** [BjornFJohansson/pydna#49]
I'm running the pydna.Assembly a number of times. I notice that when I execute my code, I get anywhere from 4-7 different assemblies, depending on the run (just running the same code over and over again using a simple bash script). Is this expected? Is the algorithm random / not deterministic?

**Comment by jlerman44** [comment]
(This is with version 2.0.3 installed using pip.  I also tried with 3.0.0 installed from source, but it runs much much slower and doesn't seem to give me expected results - i haven't debugged much though).

**Comment by BjornFJohansson** [comment]
Hi and thanks for your interest in pydna. Can you share your code and data of your example?
There was some differences in the order that assemblies were presented in earlier versions of the code, and I think these were due to the randomness of dicts. However, the same set of results should always be returned. The 3.0.0 is work in progress and I am changing how especially linear assembles are reported. Again, I would be grateful for a working example that shows your problem.

**Comment by jlerman44** [comment]
Thank you for writing pydna!  It's going to be an excellent start towards what I need to accomplish. 

The code is:

```python
import pydna
from pydna.all import read, Assembly


S106727 = read(
    """>S106727
    REDACTED_DNA_NUCLEOTIDE_SEQUENCE_1
    """.upper()
)

S106728 = read(
    """>S106728
    REDACTED_DNA_NUCLEOTIDE_SEQUENCE_2
    """.upper()
)

S106966 = read(
    """>S106966
    REDACTED_DNA_NUCLEOTIDE_SEQUENCE_3
    """.upper()
)

S106967 = read(
    """>S106967
    REDACTED_DNA_NUCLEOTIDE_SEQUENCE_4
    """.upper()
)


S106728 = S106728.reverse_complement()
S106967 = S106967.reverse_complement()

asm = Assembly( [S106727, S106728, S106966, S106967] , limit = 100)
linear_products = asm.linear_products
print('total linear products:', len(linear_products))


```

When executed, it'll produce a different number of total potential linear products each time.

I unfortunately cannot share the DNA sequences as they are confidential, but I'll try to produce a minimal non-confidential example that has the same problem. I'll also continue to review the code for anything that is potentially random. I don't see anything so far, except for color choices when exporting  to the ApE editor.  Let me know what ideas you have!

Josh

**Comment by jlerman44** [comment]
Some info about the sequences: they are big, roughly 4000 nuc each. and also here's the assembly summary in one instance:

```
total linear products: 228
Assembly:
Sequences........................: [4565] [4255] [4703] [4189]
Sequences with shared homologies.: [4565] [4255] [4255] [4565] [4703] [4703] [4189] [4189]
Homology limit (bp)..............: 100
Number of overlaps...............: 18
Nodes in graph(incl. 5' & 3')....: 15
Only terminal overlaps...........: No
Circular products................: [8904] [8834] [8832] [8681] [8679] [8134] [7981] [7911] [7909] [7756] [4884] [4720] [4718] [4245] [3961] [3889] [3797] [3795]
Linear products..................: [12512] [11997] [9482] [9481] [9348] [9347] [9323] [9321] [9261] [9259] [9189] [9187] [9127] [9125] [8948] [8914] [8813] [8810] [8712] [8664] [8661] [8661] [8660] [8660] [8657] [8657] [8652] [8603] [8590] [8589] [8559] [8558] [8556] [8556] [8553] [8537] [8537] [8526] [8526] [8523] [8503] [8499] [8498] [8465] [8465] [8454] [8441] [8437] [8436] [8434] [8433] [8428] [8425] [8425] [8424] [8422] [8400] [8399] [8398] [8398] [8395] [8395] [8368] [8367] [8365] [8364] [8338] [8336] [8336] [8333] [8333] [8303] [8302] [8302] [8301] [8298] [8275] [8275] [8267] [8267] [8266] [8264] [8264] [8237] [8216] [8213] [8213] [8212] [8211] [8208] [8206] [8205] [8205] [8204] [8202] [8202] [8175] [8175] [8144] [8144] [8140] [8140] [8113] [8106] [8088] [8078] [8050] [8044] [8044] [7988] [7984] [7982] [7888] [5440] [5436] [5339] [5338] [5335] [5334] [5280] [5278] [5276] [5274] [5239] [5238] [5234] [5233] [5218] [5216] [5214] [5212] [5180] [5178] [5176] [5174] [5118] [5116] [5114] [5112] [5080] [5078] [5075] [5073] [5018] [5016] [5013] [5011] [4804] [4801] [4704] [4702] [4701] [4700] [4669] [4666] [4644] [4642] [4603] [4600] [4582] [4580] [4568] [4568] [4565] [4544] [4542] [4517] [4514] [4513] [4510] [4507] [4482] [4480] [4480] [4467] [4464] [4445] [4442] [4417] [4416] [4415] [4414] [4413] [4412] [4412] [4411] [4409] [4407] [4406] [4380] [4357] [4355] [4353] [4351] [4345] [4344] [4316] [4315] [4313] [4311] [4310] [4308] [4306] [4295] [4293] [4291] [4289] [4257] [4255] [4253] [4251] [4250] [4244] [4195] [4193] [4193] [4191] [4188] [4157] [4155] [4152] [4150] [4095] [4093] [4090] [4088] [556] [556]

```

And another run of the same script:

```
total linear products: 203
Assembly:
Sequences........................: [4565] [4255] [4703] [4189]
Sequences with shared homologies.: [4565] [4255] [4255] [4565] [4703] [4703] [4189] [4189]
Homology limit (bp)..............: 100
Number of overlaps...............: 18
Nodes in graph(incl. 5' & 3')....: 15
Only terminal overlaps...........: No
Circular products................: [8906] [8904] [8834] [8832] [8681] [8681] [8679] [8134] [7981] [7909] [7756] [4886] [4884] [4720] [4718] [4245] [3961] [3889] [3795]
Linear products..................: [12512] [11997] [9482] [9481] [9348] [9347] [9323] [9321] [9261] [9259] [9189] [9187] [9127] [9125] [8948] [8914] [8813] [8810] [8712] [8664] [8661] [8661] [8660] [8660] [8657] [8657] [8652] [8603] [8590] [8589] [8559] [8556] [8556] [8553] [8537] [8537] [8526] [8526] [8523] [8503] [8499] [8498] [8465] [8465] [8454] [8441] [8437] [8436] [8434] [8433] [8428] [8425] [8425] [8422] [8399] [8398] [8398] [8395] [8395] [8368] [8367] [8365] [8364] [8336] [8336] [8333] [8333] [8303] [8302] [8302] [8301] [8298] [8275] [8275] [8267] [8267] [8264] [8264] [8237] [8216] [8213] [8213] [8212] [8211] [8208] [8206] [8205] [8205] [8202] [8202] [8175] [8175] [8144] [8144] [8140] [8140] [8113] [8106] [8088] [8078] [8050] [8044] [8044] [7988] [7984] [7982] [7888] [5440] [5436] [5341] [5339] [5338] [5337] [5335] [5334] [5280] [5278] [5276] [5274] [5239] [5238] [5234] [5233] [5218] [5216] [5214] [5212] [5180] [5178] [5176] [5174] [5118] [5116] [5114] [5112] [5080] [5078] [5075] [5073] [5018] [5016] [5013] [5011] [4804] [4801] [4704] [4701] [4700] [4669] [4666] [4642] [4603] [4600] [4580] [4568] [4568] [4565] [4542] [4517] [4514] [4513] [4510] [4507] [4480] [4480] [4467] [4464] [4445] [4442] [4417] [4416] [4414] [4413] [4412] [4412] [4409] [4407] [4406] [4380] [4355] [4351] [4345] [4344] [4316] [4313] [4311] [4308] [4306] [4293] [4289] [4255] [4251] [4250] [4244] [4193] [4193] [4188] [4155] [4150] [4093] [4088] [556] [556]

```

**Comment by jlerman44** [comment]
I think I may have stumbled upon a solution and I'm curious if you think it is valid:

Change line 190 in 'assembly.py' to sort on a secondary attribute in case of ties:
```python
ds.features = sorted([f for f in ds.features], key = _operator.attrgetter("location.start"))
``` 
becomes
```python
ds.features = sorted([f for f in ds.features], key = _operator.attrgetter("location.start", "location.end"))
``` 

And also lines 203-205 from
```python
            overlaps = sorted( list({f.qualifiers['chksum'][0]:f for f in dsrec.features
                                if f.type=='overlap'}.values()),
                               key = _operator.attrgetter('location.start'))
``` 
to 
```python
            overlaps = sorted( list({f.qualifiers['chksum'][0]:f for f in dsrec.features
                                if f.type=='overlap'}.values()),
                               key = _operator.attrgetter('location.start', 'location.end'))
```

After making these changes, the code runs deterministically.

**Comment by BjornFJohansson** [comment]
Thanks for you effort, Ill look into this asap.

**Comment by jlerman44** [comment]
I'm guessing it happens when there are homology blocks in a part that start at the same position, but end at different positions?

**Comment by BjornFJohansson** [comment]
Hi, I see no reason why your code could not be applied into the 2. version. I changed the code between 2.0 and 3.0 quite a bit in the assembly module.

In the 2. version, I used features to store information on the overlaps between sequences before adding them as features. This became messy for fragments that underwent several consecutive recombination reactions. 

In the new version, the nodes are stored as a list of tuples in the nodes property of each fragment. See lines 190,191 197,198 in the latest source code. These tuples contain (startpos, node_id) so these should sort in a deterministic way (lines 200, 201).

The linear assembly works differently in the new code, The algorithm reports only assemblies between the first and last sequences given as arguments. 

May I ask what is the objective in you example? Is it circular assemblies or linear? If they are linear, do you expect them to start with S106727 and end with S106967?

I have a large number of test cases for this in the test suit, but I am very interested in use cases that might break the code.

Finally, one advantage of using the new code is that features in the sequences before assembly are preserved in the final result.

**Comment by BjornFJohansson** [comment]
Regarding the ApE editor colors, instead of random colors a predefined set could be used as for example this one: http://bids.github.io/colormap/

**Comment by jlerman44** [comment]
Thanks for the detail!  Actually the objective is assembly of 2 linear fragments. The first 2 in the list and then the second 2 in the list.

**Comment by jlerman44** [comment]
But yeah now that I know more about the implementation, I’m not sure this use case quite works.

**Comment by BjornFJohansson** [comment]
I see, the S106727 is the reverse complement of S106728 and S106966 is the reverse complement of S106967 and then you expect the two resulting fragments to join? I suspect what the problem might be, I never fed the algorithm sequences and not their reverse complements, this should be taken care of by the algorithm. There is an optional [max_nodes](https://pydna.readthedocs.io/#pydna.assembly.Assembly) keyword argument that is supposed to take care of the number of sequences in the final results. It is quite easy to choke the process. 

So if I understood correctly:

S106727 and S106728 should recombine by some internal palindromic sequence into a fragment we can call "A" and S106966 and S106967 should also recombine into "B" and then A and B would recombine by a shared sequence? Or do you simply want to assemble S106727 and S106966? What are the approx lengths and locations (internal or flanking) of overlapping sequences? 

Regarding the corrections you have now, fork the pydna repo and make a pull request and I can make a new 2. bugfix release. I am interested in making this work for the new code as well as it is better in many ways.

**Comment by jlerman44** [comment]
OK I will focus development on v3.
I'm actually hoping 2 linear assembly possibilities would be found when I feed in those 4 DNA parts:
```
S106966|500
        \/
        /\
        500|S106967_rc

and

S106727|500
        \/
        /\
        500|S106728_rc
```

**Comment by jlerman44** [comment]
Also, with v2, sometimes I do notice weird self assemblies. This should be a 2 "node" assembly instead of a 3 node. 

```
8336
S106966|700
        \/
        /\
        700|S106966|500
                    \/
                    /\
                    500|S106967_rc
```

**Comment by jlerman44** [comment]
For v2 fix, I put it here: https://github.com/jlerman44/pydna/tree/deterministic_assembly 

Which branch should i send the pull request to though?   I based it on 518c4a8 since it had the same version tag as the version currently being served in PyPI here https://pypi.org/project/pydna/.

**Comment by BjornFJohansson** [comment]
Hi @jlerman44 
Sorry for the silence, I have been busy. Looked at your example and I tried to replicate what you want with short sequences. https://gist.github.com/BjornFJohansson/f7a160b8cd815cd2ff201963c2eaadb7

One big difference between the old and the new assembly code is that the new assumes that you want to 
trace assemblies that start with the first sequence and ends with the last sequence of the iterable used as argument. This way the results for linear assemblies are easier to interpret and it fitted my use case better. If there is a reason, the old functionality could be retained. 

I noticed that you fed the Assembly class four sequences but you expect them to join two by two?
If you know this beforehand, you could set up two Assembles with the two pairs?

**Comment by BjornFJohansson** [comment]
I hope this is resolved in the new version 3.0.1. 
Please reopen if not.
